### PR TITLE
create resources for SSCS

### PR DESCRIPTION
**Pull request by deads2k**
WIP while I sort out the library-go and api fixes.

Adds a simple operator that handles a basic create scenario.  It has some auto-stomping features, but not a full set.  After more common bits come out for the operator, we have surprisingly little code.

/assign @mfojtik @sttts

TODO after this
 1. configmap rotation - needed for general config updates
 2. config merging - needed to merge in user defined config
 3. secret rotation - needed to properly manage serving signer keys
 4. CA bundle writing
 5. expiry for signing keys
 6. controller healthz failure on new key present

### CODE REVIEW

File: manifests/v3.10.0/service-serving-cert-signer-controller/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        ports:
+        - containerPort: 8443
+        volumeMounts:
+        - mountPath: /var/run/configmaps/config
+          name: config
+        - mountPath: /var/run/secrets/signing-key
+          name: signing-key
+        - mountPath: /var/run/secrets/serving-cert
+          name: serving-cert
+        readinessProbe:
```
Comment (mfojtik) [inline manifests/v3.10.0/service-serving-cert-signer-controller/deployment.yaml]: can you add livenessProbe as well?

File: manifests/v3.10.0/service-serving-cert-signer-controller/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      name: service-serving-cert-signer
+      labels:
+        app: openshift-service-serving-cert-signer
+        service-serving-cert-signer: "true"
+    spec:
+      serviceAccountName: service-serving-cert-signer-sa
+      containers:
+      - name: service-serving-cert-signer-controller
+        image: ${IMAGE}
+        imagePullPolicy: IfNotPresent
```
Comment (mfojtik) [inline manifests/v3.10.0/service-serving-cert-signer-controller/deployment.yaml]: guess this should be Always in production?

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
+package operator
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/sets"
```
Comment (mfojtik) [inline pkg/operator/operator.go]: nite: move down

File: pkg/operator/operator.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+	operatorConfigClient scsclientv1alpha1.ServiceCertSignerOperatorConfigsGetter,
+	appsv1Client appsclientv1.AppsV1Interface,
+	corev1Client coreclientv1.CoreV1Interface,
+) *ServiceCertSignerOperator {
+	c := &ServiceCertSignerOperator{
+		operatorConfigClient: operatorConfigClient,
+		appsv1Client:         appsv1Client,
+		corev1Client:         corev1Client,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "WebConsoleOperator"),
```
Comment (mfojtik) [inline pkg/operator/operator.go]: s/WebConsoleOperator/ServiceServingCertSignerOperator/ (that name is horrible btw ;-)

File: pkg/operator/operator.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+	}
+	desiredVersion, err := semver.Parse(operatorConfig.Spec.Version)
+	if err != nil {
+		// TODO report failing status, we may actually attempt to do this in the "normal" error handling
+		return err
+	}
+
+	errors := []error{}
+	switch {
+	case betweenOrEmpty(currentActualVerion, "3.10.0", "3.10.1") && between(&desiredVersion, "3.10.0", "3.10.1"):
```
Comment (mfojtik) [inline pkg/operator/operator.go]: should we have these versions defined elsewhere? (const/var)

File: pkg/operator/operator.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+	}
+
+	if _, err := c.operatorConfigClient.ServiceCertSignerOperatorConfigs().UpdateStatus(operatorConfig); err != nil {
+		errors = append(errors, err)
+	}
+
+	return utilerrors.NewAggregate(errors)
+}
+
+func between(needle *semver.Version, lowerInclusive, upperExclusive string) bool {
```
Comment (mfojtik) [inline pkg/operator/operator.go]: guess this and the one below are a good library-go candidates

File: pkg/operator/operator.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	upper := semver.MustParse(upperExclusive)
+	return needle.GTE(lower) && needle.LT(upper)
+}
+
+// Run starts the serviceCertSigner and blocks until stopCh is closed.
+func (c *ServiceCertSignerOperator) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting WebConsoleOperator")
```
Comment (mfojtik) [inline pkg/operator/operator.go]: bad copy&paste

File: pkg/operator/sync_v310_00.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+	if err != nil {
+		return nil, false, err
+	}
+	configBytes, err := runtime.Encode(unstructured.UnstructuredJSONScheme, config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	requiredConfigMap := resourceread.ReadConfigMapV1OrDie(v310_00_assets.MustAsset("v3.10.0/service-serving-cert-signer-controller/cm.yaml"))
+	requiredConfigMap.Data["controller-config.yaml"] = string(configBytes)
```
Comment (mfojtik) [inline pkg/operator/sync_v310_00.go]: make the key const

File: pkg/operator/sync_v310_00.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+				generation = curr.LastGeneration
+			}
+		}
+	}
+	return resourceapply.ApplyDeployment(c.appsv1Client, required, generation, forceDeployment)
+}
+
+
+func serviceServingCertSignerName() string {
+	return fmt.Sprintf("%s@%d", "openshift-service-serving-signer", time.Now().Unix())
```
Comment (mfojtik) [inline pkg/operator/sync_v310_00.go]: you sure about using UNIX timestamp as name? can't we use generation or something?

**Comment by mfojtik** [comment]
@deads2k added some comments, mostly nits, the code looks good

File: manifests/v3.10.0/service-serving-cert-signer-controller/deployment.yaml (hunk 2 above)
Comment (deads2k) [reply manifests/v3.10.0/service-serving-cert-signer-controller/deployment.yaml]: > guess this should be Always in production?

Maybe.  Why?

File: pkg/operator/sync_v310_00.go (hunk 9 above)
Comment (deads2k) [reply pkg/operator/sync_v310_00.go]: > you sure about using UNIX timestamp as name? can't we use generation or something?
> 
>  Write

I could go UID, but generation isn't unique enough.

File: pkg/operator/sync_v310_00.go (hunk 8 above)
Comment (deads2k) [reply pkg/operator/sync_v310_00.go]: > make the key const

I really don't want other places to mirror it.  I did make it a const just to make my point though....

File: pkg/operator/sync_v310_00.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	if err != nil {
+		return nil, false, err
+	}
+	configBytes, err := runtime.Encode(unstructured.UnstructuredJSONScheme, config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	requiredConfigMap := resourceread.ReadConfigMapV1OrDie(v310_00_assets.MustAsset("v3.10.0/service-serving-cert-signer-controller/cm.yaml"))
+	const configKey = "controller-config.yaml"
```
Comment (deads2k) [inline pkg/operator/sync_v310_00.go]: @mfojtik const-ified at the proper scope!

**Pull request merged**